tc/core/execution_engine_api_rfc.h

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include <dlpack/dlpack.h>
+
+#include "tc/core/compilation_cache.h"
+#include "tc/core/tc_executor.h"
+#include "tc/core/tensor.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/tree.h"
+
+namespace tc {
+
+/**
+ * The ExecutionEngine is the entry point to using TC in C++.
+ * It provides simple backend-agnostic capabilities to:
+ *   1. construct the tree IR for parsed TCs from
+ *      a string representation and store (a shared_ptr to) it
+ *   2. infer actual tmp/output tensor shapes given input tensor shapes;
+ *   3. compile and run TCs on backend-specific TcExecutors;
+ *   4. store the TcExecutors for memoization of compilations.
+ *      The correspondance is:
+ *      1 TcExecutor <-> 1 tuple<TC function, input shapes, MappingOptions>
+ *
+ * ExecutionEngine is templated by the Backend type.
+ * For each backend, the specific Backend type lives in
+ *   backend/backed_tc_executor.h and declares all the required dependent
+ *   **concrete** types.
+ * For example:
+ *   CudaBackend is declared in core/cuda/cuda_tc_executor.h
+ *
+ * struct CudaBackend {
+ *   using ExecutorType = CudaTcExecutor;
+ *   using MappingOptionsType = CudaMappingOptions;
+ *   using ProtobufType = CudaCacheProto;
+ *   using CacheType = CudaCache;
+ *   using CacheEntryType = CudaCacheEntry;
+ *   using CacheKeyType = typename CacheEntryType::Key;
+ *   using CacheValueType = typename CacheEntryType::Values;
+ *   using RTCFunctionType = CudaRTCFunction;
+ * };
+ *
+ * Sketching usage resembles:
+ *   std::string someTc = "...";
+ *   ExecutionEngine<CudaBackend> engine;
+ *   engine.define(someTc)
+ *   auto handle = engine.compile(someTc, inputs, mappingOptions)
+ *   auto cpuKernelTimings = engine.run(handle, inputs, outputs, true);
+ *   // alternatively:
+ *   // auto kernelTiming = engine.uncheckedRun(handle, inputs, outputs, true);
+ */
+template <typename Backend>
+class ExecutionEngine {
+ public:
+  using BackendType = Backend;
+  using ExecutorType = typename BackendType::ExecutorType;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+
+  ExecutionEngine() = default;
+
+  /// Parse TC definitions provided as string, store parsed trees internally.
+  void define(const std::string& language);
+
+  /// Store the provided parsed trees internally.
+  void define(const std::vector<lang::TreeRef>& treeRefs);
+
+  /// For a given TC kernel name, compute the shapes of the output tensors
+  /// provided the shapes of the input TC tensors.  The caller can use the
+  /// computed shapes to allocate memory for outputs.
+  std::vector<TensorInfo> inferOutputTensorInfo(
```

Comment:
`TensorInfo` is never defined.  It is a struct from the autotuner IIRC, not sure we want to carry it here.  Why not `DLTensor` ?

Reply:
One problem with `DLTensor` is memory management: it's using raw pointers to store strides/shapes.